### PR TITLE
feat(FEC-7551): handling keyboard a11y

### PR DESCRIPTION
**Pull request by RoyBregman**
User should be able to navigate menus using the TAB and the UP/DOWN arrows

### CODE REVIEW

File: src/components/language/language.js [hunk 1]
```diff
@@ -245,11 +231,36 @@ class LanguageControl extends BaseComponent {
...
+    return (
+      <div
+        tabIndex="-1"
+        ref={el => {
+          if (props.pushRef) {
+            props.pushRef(el);
+          }
+        }}
+        className={style.smartContainerItem}
+        onSelect={props.onSelect}>
```
Comment (RoyBregman) [inline src/components/language/language.js]: check if needed (onSelect)

File: src/components/menu/menu.js [hunk 2]
```diff
@@ -215,23 +201,58 @@ class Menu extends Component {
...
+        }}
+        className={[style.dropdownMenu, ...this.state.position].join(' ')}>
         {props.options.map((o, index) => (
-          <div
-            tabIndex=""
+          <KeyboardAccessibleMenuItem
+            setFirstFocusedElement={props.setFirstFocusedElement}
+            addAccessibleChild={props.addAccessibleChild}
+            isSelected={this.isSelected}
+            onSelect={this.onSelect.bind(this)}
```
Comment (RoyBregman) [inline src/components/menu/menu.js]: change to arrow function

File: src/components/menu/menu.js [hunk 3]
```diff
@@ -215,23 +201,58 @@ class Menu extends Component {
...
+  }
+
+  render(props: any): React$Element<any> {
+    return (
+      <div
+        tabIndex="-1"
+        ref={se => {
+          this._selectedElement = props.isSelected(props.data) ? se : this._selectedElement;
+          props.setFirstFocusedElement(this._selectedElement);
+          if (!this._mounted) {
```
Comment (RoyBregman) [inline src/components/menu/menu.js]: check for better pattern to add it only onMount

File: src/ui-presets/playback.js [hunk 4]
```diff
@@ -41,6 +41,7 @@ export function playbackUI(props: any): React$Element<any> {
     <div className={style.playbackGuiWWrapper}>
       <KeyboardControl player={props.player} config={props.config} />
       <Loading player={props.player} />
+      <PictureInPictureOverlay player={props.player} />
```
Comment (RoyBregman) [inline src/ui-presets/playback.js]: check if it better inside player gui

File: src/utils/popup-item-keyboard-accessibility.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    /**
+     * render component
+     *
+     * @param {*} props - component props
+     * @returns {React$Element} - component element
+     * @memberof HOC
+     */
+    render(props: any): React$Element<any> | void {
+      return h(
+        'div',
```
Comment (RoyBregman) [inline src/utils/popup-item-keyboard-accessibility.js]: use jsx

File: src/components/cvaa-overlay/cvaa-overlay.js [hunk 6]
```diff
@@ -273,73 +161,114 @@ class CVAAOverlay extends BaseComponent {
...
    */
-  renderCustomCaptionsState(props: any): React$Element<any> {
-    const fontFamily = this.props.player.TextStyle.FontFamily;
-    const edgeStyles = this.props.player.TextStyle.EdgeStyles;
+  render(props: any): React$Element<any> {
+    return (
+      <Overlay open onClose={() => props.onClose()} type="cvaa">
+        {this.state.activeWindow === cvaaOverlayState.Main ? (
+          <KeyboardAccessibleMainWindow
+            tabbable="true"
```
Comment (yairans) [inline src/components/cvaa-overlay/cvaa-overlay.js]: should not be `true`? or just `tabbable`

File: src/components/dropdown/dropdown.js [hunk 7]
```diff
@@ -138,10 +143,15 @@ class DropDown extends Component {
         className={this.state.dropMenuActive ? [style.dropdown, style.active].join(' ') : style.dropdown}
         ref={el => (this._el = el)}>
         <div
-          tabIndex="0"
-          ref={el => (this._dropdownButton = el)}
+          tabIndex={props.tabbable ? '0' : -1}
```
Comment (yairans) [inline src/components/dropdown/dropdown.js]: `tabIndex` is `string` or `number`?

File: src/components/menu/menu.js [hunk 8]
```diff
@@ -215,23 +201,58 @@ class Menu extends Component {
...
+            props.addAccessibleChild(se);
+          }
+        }}
+        className={props.isSelected(props.data) ? [style.dropdownMenuItem, style.active].join(' ') : style.dropdownMenuItem}
+        onClick={() => this.onSelect(props)}>
+        <span>{props.data.label}</span>
+        <span className={style.menuIconContainer} style={`opacity: ${props.isSelected(props.data) ? 1 : 0}`}>
```
Comment (yairans) [inline src/components/menu/menu.js]: handle the  opacity by class

File: src/components/smart-container/smart-container.js [hunk 9]
```diff
@@ -82,16 +84,30 @@ class SmartContainer extends Component {
...
+          ...this.props
+        );
+      }
+    });
+    return children;
+  }
 }
 
-export {SmartContainer};
+const keyboardAccessibleSmartContainer = popupWithKeyboardA11y(SmartContainer);
```
Comment (yairans) [inline src/components/smart-container/smart-container.js]: can't use `@with...` pattern here?

File: src/utils/popup-item-keyboard-accessibility.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
+//@flow
+import {BaseComponent} from '../components/base';
+import {h} from 'preact';
+import {KeyMap} from '../utils/key-map';
+
+/**
+ * @param {BaseComponent} WrappedComponent - The popup item component to implement keyboard accessibility
+ * @returns {BaseComponent} - HOC that handles animation
```
Comment (yairans) [inline src/utils/popup-item-keyboard-accessibility.js]: fix the desc

File: src/utils/popup-keyboard-accessibility.js [hunk 11]
```diff
@@ -0,0 +1,111 @@
+//@flow
+import {BaseComponent} from '../components/base';
+import {h} from 'preact';
+import {KeyMap} from '../utils/key-map';
+
+/**
+ * @param {BaseComponent} WrappedComponent - The popup component to implement keyboard accessibility
+ * @returns {BaseComponent} - HOC that handles animation
```
Comment (yairans) [inline src/utils/popup-keyboard-accessibility.js]: fix the desc

File: src/utils/popup-keyboard-accessibility.js [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+export const popupWithKeyboardA11y: Function = (WrappedComponent: BaseComponent): typeof BaseComponent =>
+  class extends BaseComponent {
+    _firstFocusedElement: HTMLElement;
+    _accessibleChildren: Array<HTMLElement> = [];
+    _activeElement: ?HTMLElement;
+    _previouslyActiveElement: ?HTMLElement;
+    _element: ?HTMLElement;
+
+    /**
+     * after component mounted, listen to events
```
Comment (yairans) [inline src/utils/popup-keyboard-accessibility.js]: fix

File: src/components/cast-on-tv/cast-before-play.js [hunk 13]
```diff
@@ -88,8 +89,16 @@ class CastBeforePlay extends BaseComponent {
       }
       return (
         <div>
-          <div className={rootStyle.join(' ')} onClick={() => this.onClick()}>
-            <a className={[style.btn, style.btnDarkTransparent, style.castOnTvButton].join(' ')}>
+          <div className={rootStyle.join(' ')}>
+            <a
+              tabIndex="0"
```
Comment (yairans) [inline src/components/cast-on-tv/cast-before-play.js]: `tabIndex` is string or number?

File: src/components/cast/cast.js [hunk 14]
```diff
@@ -60,10 +61,17 @@ class CastControl extends BaseComponent {
...
+            if (e.keyCode === KeyMap.ENTER) {
+              this.props.updateBackdropVisibility(true);
+              this.player.startCasting().catch(() => this.props.updateBackdropVisibility(false));
+            }
+          }
         },
         h('google-cast-launcher', {
-          class: style.castButton
+          class: [style.castButton].join(' '),
+          tabIndex: 0
```
Comment (yairans) [inline src/components/cast/cast.js]: see my question above

File: src/components/smart-container/smart-container.js (hunk 9 above)
Comment (yairans) [reply src/components/smart-container/smart-container.js]: see https://github.com/kaltura/playkit-js-ui/pull/410

File: src/utils/popup-item-keyboard-accessibility.js (hunk 5 above)
Comment (yairans) [reply src/utils/popup-item-keyboard-accessibility.js]: see example here https://github.com/kaltura/playkit-js-ui/pull/410

File: src/components/cast-on-tv/cast-before-play.js [hunk 15]
```diff
@@ -88,8 +89,16 @@ class CastBeforePlay extends BaseComponent {
       }
       return (
         <div>
-          <div className={rootStyle.join(' ')} onClick={() => this.onClick()}>
-            <a className={[style.btn, style.btnDarkTransparent, style.castOnTvButton].join(' ')}>
+          <div className={rootStyle.join(' ')}>
+            <a
```
[the remark this answers is not in the document]
Comment (RoyBregman) [reply src/components/cast-on-tv/cast-before-play.js]: done

File: src/components/cast-on-tv/cast-before-play.js (hunk 13 above)
Comment (RoyBregman) [reply src/components/cast-on-tv/cast-before-play.js]: html attributes are strings

File: src/components/cast/cast.js (hunk 14 above)
Comment (RoyBregman) [reply src/components/cast/cast.js]: refactored entire code to jsx

File: src/components/smart-container/smart-container.js (hunk 9 above)
Comment (RoyBregman) [reply src/components/smart-container/smart-container.js]: AMAZING!!!! done

File: src/utils/popup-keyboard-accessibility.js (hunk 11 above)
Comment (RoyBregman) [reply src/utils/popup-keyboard-accessibility.js]: see now

File: src/utils/popup-keyboard-accessibility.js (hunk 12 above)
Comment (RoyBregman) [reply src/utils/popup-keyboard-accessibility.js]: fix what? :)

File: src/components/smart-container/smart-container.js [hunk 16]
```diff
@@ -76,14 +75,19 @@ class SmartContainer extends Component {
...
         <Overlay open onClose={() => props.onClose()}>
           <div className={style.title}>{props.title}</div>
           {props.children}
         </Overlay>
       </Portal>
     ) : (
-      <div tabIndex="-1" className={[style.smartContainer, style.top, style.left].join(' ')}>
+      <div
+        onKeyDown={e => {
+          props.handleKeyDown(e);
```
Comment (RoyBregman) [inline src/components/smart-container/smart-container.js]: check to remove

**Pull request merged**